cpp/src/reductions/histogram.cu

```diff
@@ -0,0 +1,298 @@
+
+  return std::make_unique<cudf::list_scalar>(
+    std::move(*output_structs.release()), true, stream, mr);
+}
+
+}  // namespace
+
+std::pair<rmm::device_uvector<size_type>, std::unique_ptr<column>> table_histogram(
+  table_view const& input,
+  std::optional<column_view> const& partial_counts,
+  data_type const output_dtype,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(cudf::is_integral(output_dtype),
+               "The output count type of histogram aggregation must be an integral type.");
+
+  auto map = cudf::detail::hash_map_type{
+    compute_hash_table_size(input.num_rows()),
+    cuco::empty_key{-1},
+    cuco::empty_value{std::numeric_limits<size_type>::min()},
+    cudf::detail::hash_table_allocator_type{default_allocator<char>{}, stream},
+    stream.value()};
+
+  auto const preprocessed_input =
+    cudf::experimental::row::hash::preprocessed_table::create(input, stream);
+  auto const has_nulls          = nullate::DYNAMIC{cudf::has_nested_nulls(input)};
+  auto const has_nested_columns = cudf::detail::has_nested_columns(input);
+
+  auto const row_hasher = cudf::experimental::row::hash::row_hasher(preprocessed_input);
+  auto const key_hasher = row_hasher.device_hasher(has_nulls);
+  auto const row_comp   = cudf::experimental::row::equality::self_comparator(preprocessed_input);
+
+  auto const pair_iter = cudf::detail::make_counting_transform_iterator(
+    size_type{0}, [] __device__(size_type const i) { return cuco::make_pair(i, i); });
+
+  // Always compare NaNs as equal.
+  using nan_equal_comparator =
+    cudf::experimental::row::equality::nan_equal_physical_equality_comparator;
+  auto const value_comp = nan_equal_comparator{};
+
+  if (has_nested_columns) {
+    auto const key_equal = row_comp.equal_to<true>(has_nulls, null_equality::EQUAL, value_comp);
+    map.insert(pair_iter, pair_iter + input.num_rows(), key_hasher, key_equal, stream.value());
+  } else {
+    auto const key_equal = row_comp.equal_to<false>(has_nulls, null_equality::EQUAL, value_comp);
+    map.insert(pair_iter, pair_iter + input.num_rows(), key_hasher, key_equal, stream.value());
+  }
+
+  // Gather the indices of distinct rows.
+  auto distinct_indices = rmm::device_uvector<size_type>(
+    static_cast<size_type>(map.get_size()), stream, rmm::mr::get_current_device_resource());
+
+  // Store the number of occurrences of each distinct row.
+  auto distinct_counts = make_numeric_column(
+    output_dtype, static_cast<size_type>(map.get_size()), mask_state::UNALLOCATED, stream, mr);
+
+  type_dispatcher(output_dtype,
+                  histogram_dispatcher{},
+                  map,
+                  std::move(preprocessed_input),
+                  input.num_rows(),
+                  has_nulls,
+                  has_nested_columns,
+                  distinct_indices.begin(),
+                  distinct_counts->mutable_view(),
+                  partial_counts,
+                  stream);
```

Comment:
Using `type_dispatcher` for error check is overkill.

Reply:
But `type_dispatcher` allows extending the output (count) type. Currently we support only `int64_t` but we may have more if needed. I'm open to just using only one output count type and removing `type_dispatcher`.

Reply:
IIUC, a dispatcher is needed if there are significantly different semantics based on the type, e.g. between nested and flat data. Are we expecting to support types other than integers in this case?

Reply:
Count types must be integer. Currently int64 is used, but I'm not sure if int32 will be needed in other applications thus I intentionally leave this open for extension.

Reply:
Right, my point is that there is no benefit in distinguishing the code path between int32 and int64 if we know the count is always an integer.

Reply:
I thought more about this. Probably `int64` should be the (default) only option, so the user won't have to make choice between int32 and int64 and won't have potential overflow bug with int32.